<b>############ INDELWALD - HYBRID INDEL CALLING ##############</b>

<p>## Version 1.0 ##</p>
<p>## Last Update - 04/08/2015 ##</p>
<p>## Developped as part a Computational Biology MPhil Internship ##</p>
<p>## University of Cambridge, UK ##</p>
<p>## mrs72 / Maximilian Stammnitz ##</p>

<p>This small framework consists of a few components for analysing DFTD indels:</p>
1. An R-file containing all functions for post-processing and displaying Pindel and Platypus calls
2. An R-file example, guiding through the main options of output lists and plots
3. A small Platypus output example VCF
4. A small Pindel output example VCF
5. A table with > 36,000 sorted supercontigs of the Devil Reference Genome 7.1, their respective sizes and chromosomal coordinates
6. A table with all biomaRt-translated ENSEMBL gene and transcript IDs
7. A table comprising all annotated devil exons and introns, chromosomal coordinate, strand orientation, ENSEMBL gene and transcript ID
8. A CSV-table with all annotated COSMIC cancer gene consensus genes and their characteristics

<b>############ 14-08-2015: UPDATE ##############</b>
<b>############ FILE UPLOAD IN VERY SHORT. ##############</b>

Portions of the raw input of our toy example, and associated file/function names of Indelwald, comprise strictly confidential informations on a recently submitted article on the origin of these samples (Pye et al.). Our group was hoping until the last that this paper would be released before the deadline of the Indelwald-associated thesis ('Indels in Tasmanian Devils'), but unfortunately it is ongoingly reviewed. In order to share the code nevertheless, a careful renaming (and some revision) has been undertaken in the last two days. Non-conflicting code will be released later today, on <b> Friday, 14th August</b>.
